src/olympia/files/utils.py

```
@@ -837,3 +842,89 @@ def _update_version_in_json_manifest(content, new_version_number):
+        xpi_path = file_obj.name
+    else:
+        raise ValueError('A path, file-like or FileUpload object is required')
+
+    messages = {}
+
+    try:
+        with zipfile.ZipFile(xpi_path, 'r') as source:
+            file_list = source.namelist()
+
+            locales = {name[8:].strip('/') for name in file_list
+                       if name.startswith('_locales')}
+
+            for locale in locales:
+                corrected_locale = find_language(locale)
+
+                # Filter out languages we don't support.
+                if not corrected_locale:
+                    continue
+
+                fname = '_locales/{0}/messages.json'.format(locale)
+
+                try:
+                    messages[corrected_locale] = json.loads(source.read(fname))
+                except KeyError:
+                    # thrown by `source.read` usually means the file doesn't
+                    # exist for some reason, we fail silently
+                    continue
+    except IOError:
+        pass
+
+    return messages
+
+
+def resolve_i18n_message(
+        message, file_obj=None, locale=None, default_locale=None,
+        messages=None):
+    """Resolve a translatable string in an add-on.
+
+    This matches ``__MSG_extensionName__`` like names and returns the correct
+    translation for `locale`.
+
+    :param file_obj: A `files.File` object or string pointing to the XPI
+                     that should be used for extraction.
+    :param locale: The locale to fetch the translation for, If ``None``
+                   (default) ``settings.LANGUAGE_CODE`` is used.
+    :param messages: A dictionary of messages, e.g the return value
+                     of `extract_translations`.
+    """
+    assert file_obj is not None or messages is not None
+
+    match = MSG_RE.match(message)
+
+    if match is None:
+        return message
+
+    msgid = match.group('msgid')
+
+    if locale is None:
+        locale = settings.LANGUAGE_CODE
```

I don't think you need this. If `locale` is None, something weird is going on and we should fail loudly.